Convert mouselistener plugins to application plugins.

malueck opened the pull request:
This pull request is supposed to fix https://github.com/JetBrains/MPS-extensions/issues/1168.

* The plugins for the editor extensions were converted from project plugin to application plugin and so they are instantiated only once per editor component, instead of once per opened project. Especially when they are registering mouse listeners, this did lead to the same listener being registered multiple times.
Change performed in:
  * com.mbeddr.mpsutil.intentions.runtime.plugin (MyIntentionMenuProducer was set in every editor multiple times)
  * de.itemis.mps.selection.runtime.plugin (multiple mouse listeners in each component)
  * de.itemis.mps.mouselistener.runtime.plugin (same)

* Furthermore, `com.mbeddr.mpsutil.intentions` was converted from the deprecated `EditorComponentCreationListener` to an `EditorExtension`, similar to https://github.com/JetBrains/MPS-extensions/commit/84e8c430b9b21bba9d7c69e3e1587232d5011cd4.

* The pagination plugin is now migrated to a plugin solution like the other plugins, instead of a plugin language aspect.

* A new solution `de.itemis.mps.editor.collapsible.sandbox` was created to test these changes, it looks like collapsible did not have a sandbox before.

* In the intentions demo language, fixed a typo `ReadyOnly` -> `ReadOnly` that was in multiple editors and intentions.

*  In `QueryListChecker`, an exception was fixed when an editor cell has a null root cell:

<details>
<summary>Stacktrace</summary>

```
java.lang.IllegalArgumentException: Argument for @NotNull parameter 'start' of jetbrains/mps/openapi/editor/cells/CellTraversalUtil.iterateTree must not be null
	at jetbrains.mps.openapi.editor.cells.CellTraversalUtil.$$$reportNull$$$0(CellTraversalUtil.java)
	at jetbrains.mps.openapi.editor.cells.CellTraversalUtil.iterateTree(CellTraversalUtil.java)
	at c.m.m.e.querylist.runtime//com.mbeddr.mpsutil.editor.querylist.runtime.plugin.QueryListChecker.collectNodesToCheck(QueryListChecker.java:107)
	at c.m.m.e.querylist.runtime//com.mbeddr.mpsutil.editor.querylist.runtime.plugin.QueryListChecker.update(QueryListChecker.java:74)
	at jetbrains.mps.nodeEditor.checking.EditorChecker.update(EditorChecker.java:70)
	at jetbrains.mps.nodeEditor.highlighter.HighlighterUpdateSession$HighlighterReadAction.perform(HighlighterUpdateSession.java:317)
	at jetbrains.mps.nodeEditor.highlighter.EditorCheckerWrapper.withChecker(EditorCheckerWrapper.java:69)
	at jetbrains.mps.nodeEditor.highlighter.HighlighterUpdateSession$HighlighterReadAction.execute(HighlighterUpdateSession.java:311)
	at jetbrains.mps.smodel.CancellableReadAction.run(CancellableReadAction.java:86)
	at jetbrains.mps.smodel.ActionDispatcher.dispatch(ActionDispatcher.java:98)
	at jetbrains.mps.smodel.ActionDispatcher.lambda$wrap$0(ActionDispatcher.java:135)
	at jetbrains.mps.smodel.LockRunnable.run(LockRunnable.java:60)
	at jetbrains.mps.smodel.PlatformCancelBlock.run(PlatformCancelBlock.java:22)
	at com.intellij.openapi.application.impl.RwLockHolder.runReadAction(RwLockHolder.kt:254)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:840)
	at jetbrains.mps.smodel.WorkbenchModelAccess.runReadAction(WorkbenchModelAccess.java:95)
	at jetbrains.mps.smodel.ModelAccessBase.runReadAction(ModelAccessBase.java:66)
	at jetbrains.mps.nodeEditor.highlighter.HighlighterUpdateSession.updateEditor(HighlighterUpdateSession.java:218)
	at jetbrains.mps.nodeEditor.highlighter.HighlighterUpdateSession.updateEditorComponent(HighlighterUpdateSession.java:185)
	at jetbrains.mps.nodeEditor.highlighter.HighlighterUpdateSession.lambda$doUpdate$0(HighlighterUpdateSession.java:100)
	at jetbrains.mps.typechecking.backend.TypecheckingSessionHandler.runWithSession(TypecheckingSessionHandler.java:76)
	at jetbrains.mps.nodeEditor.highlighter.HighlighterUpdateSession.doUpdate(HighlighterUpdateSession.java:98)
	at jetbrains.mps.nodeEditor.highlighter.HighlighterUpdateSession.update(HighlighterUpdateSession.java:257)
	at jetbrains.mps.nodeEditor.Highlighter$ScheduleHighlighterUpdate.update(Highlighter.java:490)
	at jetbrains.mps.nodeEditor.Highlighter$ScheduleHighlighterUpdate.lambda$run$0(Highlighter.java:471)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

</details>






<strike>* Changed `MyIntentionMenuProducer` so that it no longer overwrites `myShowIntentionsAction` and `myLightBulb` per reflection. Perhaps this fixes also https://github.com/JetBrains/MPS-extensions/issues/944 and https://github.com/JetBrains/MPS-extensions/issues/1086, but anyway the reflection-based approach does not work anymore since MPS introduced the `IntentionMenuProducer` interface that is now subclassed by `MyIntentionMenuProducer`.
  * The "grouping" feature of the plugin works and does not require reflection, because we override just [getIntentionsGroup](http://127.0.0.1:63320/node?ref=r%3A990d360b-3ac3-45fa-8ed3-0bbf017bba84%28com.mbeddr.mpsutil.intentions.runtime%29%2F2659720833134824971) for this, which is called by the base implementation.
  * What does NOT work is the special handling of intentions that are shown even in read-only editors, which is not supported in MPS without the plugin. Such intentions will not show up and the bulb menu will not be displayed.
  * The reason is that both the AbstractAction `myShowIntentionsAction` and the LightBulbMenu `myLightBulb` which are created in the superclass are immediately registered as actions in the node editor. Any further overwriting of these fields (by reflection or otherwise) has no effect, and methods like `checkAndShowMenu()` in the plugin are never called.
  * A partial workaround would be to search the registered keyStroke until we find the registered AbstractAction, unregister it and register our own action, but this is not possible for the lightbulb icon logic.
  * The nonfunctional code was commented out and the above explanation added as a comment. It can be either commented in again with additional workarounds, or replaced by a clean solution when MPS allows to override methods like `checkAndShowMenu` as well and removes similar hardcoded checks like https://github.com/JetBrains/MPS/blob/master/editor/editor-runtime/source/jetbrains/mps/nodeEditor/IntentionsSupport.java#L311</strike>

malueck commented:
Dropped changes to MyIntentionMenuProducer again because tests failed, will investigate this later separately from this pull request.

alexanderpann reviewed (changes requested):
Thanks a lot for the changes, they look great! Just one small change:
- de.itemis.mps.editor.collapsible.sandbox should be added to the tests build script so that it can be checked for pending migrations. It should also be located in the folder code/widgets instead of code/solutions.

malueck commented:
> Thanks a lot for the changes, they look great! Just one small change:
> 
>     * de.itemis.mps.editor.collapsible.sandbox should be added to the tests build script so that it can be checked for pending migrations. It should also be located in the folder code/widgets instead of code/solutions.

Added the solution to the test script, and also moved the solution now. I only created it from within the project view and didn't notice it was in the wrong folder first, thanks for checking.

alexanderpann reviewed (approved):
Yeah, it's always the same issue when you have a different naming scheme than the standard one from MPS.
